discover-providers.js

```diff
@@ -73,13 +77,63 @@ function createJsonFileDiscoverProvider(hostsFile) {
 }
 
 /**
+ * Wrap a innerDiscoverProvider with a retrying mechanism.
+ *
+ * @param {Object} opts the options used as a Backoff policy {@see Backoff}
+ * @param {Object} [opts.backoff] a mocked backoff for testing
+ * @param {DiscoverProvider} innerDiscoverProvider The discover provider to wrap
+ *
+ * @returns {DiscoverProvider} The retrying discover provider.
+ */
+function retryDiscoverProvider(opts, innerDiscoverProvider) {
+    if (!opts) {
+        return innerDiscoverProvider;
+    }
+
+    return function retryingDiscoverProvider(callback) {
+        var policy;
+        var defaultPolicy = {
+            minDelay: 500,
+            maxDelay: 15000,
+            timeout: 60000
+        };
+
+        if (opts === true) {
+            policy = defaultPolicy;
+        } else {
+            policy = _.defaults({}, opts, defaultPolicy);
+        }
+        var backoff = opts.backoff || new Backoff(policy);
+        var lastError = null;
+
+        attempt(null);
+
+        function attempt(fail) {
+            if (fail) {
+                return setImmediate(callback, lastError);
```

Comment:
After giving this another thought, it might make more sense to always pass an error to the callback. This is fairly easy by doing `lastError || fail`. What do you think?